modules/javafx.graphics/src/main/java/com/sun/javafx/css/FixedCapacitySet.java

```diff
@@ -0,0 +1,582 @@
+ *
+ * <p>The fall back set implementation for large sets uses open addressing. It is
+ * only lightly optimized as the expectation is that it will see little to no use.
+ * It is still preferred over {@code HashSet} due to its low memory foot print, and
+ * faster iteration.
+ *
+ * @param <T> the element type
+ */
+public sealed abstract class FixedCapacitySet<T> extends AbstractSet<T> {
+    private static final FixedCapacitySet<?> EMPTY;
+
+    static {
+        EMPTY = new Single<>();
+
+        EMPTY.freeze();
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T> FixedCapacitySet<T> empty() {
+        return (FixedCapacitySet<T>) EMPTY;
+    }
+
+    /**
+     * Creates a new {@link FixedCapacitySet} with the given maximum capacity.
+     * If the capacity is exceeded, fixed capacity sets do not grow, but instead
+     * throw an {@link IllegalStateException}.
+     *
+     * @param <T> the element type
+     * @param maximumCapacity the maximum possible number of elements the set can hold, cannot be negative
+     * @return a new empty set, never {@code null}
+     */
+    public static <T> FixedCapacitySet<T> of(int maximumCapacity) {
+        return maximumCapacity == 0 ? empty()
+             : maximumCapacity == 1 ? new Single<>()
+             : maximumCapacity == 2 ? new Duo<>()
+             : maximumCapacity < 10 ? new Hashless<>(maximumCapacity)  // will reject negative values
+                                    : new OpenAddressed<>(maximumCapacity);
+    }
+
+    private boolean frozen;
+
+    /**
+     * Checks if the given collection contains all elements
+     * of this collection. This is the same as {@link #containsAll(Collection)}
+     * with the source and target reversed, ie. {@code "a.containsAll(b)"} is equivalent
+     * to {@code "b.isSuperSetOf(a)"}.
+     *
+     * <p>If the given collection is small, or has good {@link #contains(Object)}
+     * performance, using this inverse function avoids creating an {@link Iterator}
+     * for this collection.
+     *
+     * @param c a collection to check, cannot be {@code null}
+     * @return {@code true} if the given collection contains all elements of this
+     *   collection, otherwise {@code false}
+     */
+    public abstract boolean isSuperSetOf(Collection<?> c);
+
+    /**
+     * Freezes this collection, turning it permanently read-only. After freezing,
+     * any method that would modify the collection will instead throw
+     * {@link UnsupportedOperationException}.
+     *
+     * <p>This can be used to avoid wrapping the collection with an unmodifiable
+     * collection or making a read-only copy.
+     */
+    public final void freeze() {
+        this.frozen = true;
+    }
+
+    /**
+     * Checks if the set is allowed to be mutated, and throws an
+     * {@link UnsupportedOperationException} otherwise.
+     */
+    protected final void ensureNotFrozen() {
+        if (frozen) {
+            throw new UnsupportedOperationException();
```

Comment:
I'm not against adding an explanation, but it's not needed as this is part of the `Set` contract (sets that can't be modified are specified to throw `UnsupportedOperationException`), for example for `add` in `Set`:

     * @throws UnsupportedOperationException if the {@code add} operation
     *         is not supported by this set

The concept of "freezing" the set is not visible to any users, it's just there to avoid having to make a final copy of the set. It's modifiable by the code that created it, but if they freeze it before exposing it to the outside world, then for all intents and purposes it is a read only set for anyone else.

Reply:
But it is visible: **public** `void freeze();`

Reply:
I'm not sure there is a need, but if text is added, it should be something simple like "Unmodifiable set". I would _not_ use the word "freeze" since there is no such concept in the API. From the POV of the application, the set is unmodifiable.